### PR TITLE
reshape_diagnosis up

### PR DESCRIPTION
**Pull request by acoppock**
This function reshapes a diagnosis to return a data frame that is not tidy, but is easier to read!

### CODE REVIEW

**Comment by acoppock** [comment]
@jaspercooper and @macartan -- I think this function will do nicely!

**Comment by jaspercooper** [comment]
looks really great to me. rbind_disjoint is a really nice implementation for when you're only bootstrapping from some diagnosands, or have appended other params to your diagnosands_df. really cool. will deprecate from design lib

File: R/diagnosis_helper_functions.R [hunk 1]
```diff
@@ -85,3 +85,86 @@ print.summary.diagnosis <- function(x, ...) {
...
+reshape_diagnosis <- function(diagnosis, digits = 2){
+
+  diagnosands_df <- diagnosis$diagnosands
+  diagnosands_df_names <- names(diagnosands_df)
+
+  group_names <- diagnosands_df_names[1:3]
+  value_names <- diagnosands_df_names[-(1:3)]
+
+
+  diagnosands_only_names <- value_names[which(substr(value_names, start = 1, stop = 3) != "se(")]
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: `grep('^se(', value_names, value=TRUE)`

File: R/diagnosis_helper_functions.R [hunk 2]
```diff
@@ -85,3 +85,86 @@ print.summary.diagnosis <- function(x, ...) {
...
+  diagnosands_only_df <- diagnosands_df[,c(group_names, diagnosands_only_names), drop = FALSE]
+
+  diagnosands_only_df <- cbind(diagnosands_only_df[, group_names, drop = FALSE],
+                               data.frame(statistic = "Diagnosand Estimate"),
+                               data.frame(lapply(
+                                 diagnosands_only_df[, diagnosands_only_names, drop = FALSE], format_num, digits = digits)))
+
+  if(sum(grepl(value_names, pattern = "se\\(")) == 0) {
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: `any()`

File: R/diagnosis_helper_functions.R [hunk 3]
```diff
@@ -85,3 +85,86 @@ print.summary.diagnosis <- function(x, ...) {
...
+    return(diagnosands_only_df)
+  }
+
+  se_only_names <- value_names[which(substr(value_names, start = 1, stop = 3) == "se(")]
```
Comment (nfultz) [inline R/diagnosis_helper_functions.R]: `grep`

File: R/utilities.R [hunk 4]
```diff
@@ -139,3 +139,17 @@ rbind_disjoint <- function(list_of_df, infill=NA) {
 `%i%` <- intersect
 
 `%||%` <- function(e1, e2) if(is.null(e1)) e2 else e1
+
+
+add_parens <- function(x, digits = 3) {
+  x <- as.numeric(x)
+  return_vec <- paste0("(", sprintf(paste0("%.", digits, "f"), x), ")")
```
Comment (nfultz) [inline R/utilities.R]: `sprintf("(%s)", format_num(x, digits))`

File: tests/testthat/test-diagnosands.R [hunk 5]
```diff
@@ -14,13 +14,30 @@ pate_estimator <- declare_estimator(Y ~ Z, estimand = pate, label = "test")
...
+  pate_estimator_2 <- declare_estimator(Y ~ Z, estimand = pate, label = "test_2")
+
+  my_design <- declare_design(my_population,
+                              my_potential_outcomes, pate,
+                              my_assignment,
+                              reveal_outcomes,
+                              pate_estimator,pate_estimator_2)
+
+  diagnosis <- diagnose_design(my_design, sims = 10)
+  reshape_diagnosis(diagnosis)
```
Comment (nfultz) [inline tests/testthat/test-diagnosands.R]: Please add a expectation here that actually checks the output (number of columns, number of rows, column names, etc). set.seed if necessary.

**Review by nfultz — CHANGES_REQUESTED** [review]

File: R/diagnosis_helper_functions.R (hunk 1 above)
Comment (acoppock) [reply R/diagnosis_helper_functions.R]: I think I'm not quite grokking this grep. I get:

```
grep('^se(', value_names, value=TRUE)
Error in grep("^se(", value_names, value = TRUE) : 
  invalid regular expression '^se(', reason 'Missing ')''
```

Comment (acoppock) [reply R/diagnosis_helper_functions.R]: got it `grep('^se\\(', value_names, value=TRUE)`

**Comment by acoppock** [comment]
Thanks @nfultz those changes in.

**Review by nfultz — APPROVED** [review]

**Review by graemeblair — APPROVED** [review]
LGTM

**Comment by macartan** [comment]
This version of reshape_diagnosis has issues with multiple designs.

```
> designs   <- expand_design(simple_two_arm_designer, N = 6:7)
> diagnosis <- diagnose_design(designs, sims = 10)
> reshaped  <- reshape_diagnosis(diagnosis)
Warning message:
In FUN(X[[i]], ...) : NAs introduced by coercion
> reshaped
  design_ID estimator_label coefficient           statistic estimand_label   bias   rmse  power coverage
1         1    my_estimator           Z Diagnosand Estimate             NA  -0.05   0.82   0.00     0.90
2         2    my_estimator           Z Diagnosand Estimate             NA  -0.21   0.65   0.10     0.90
3         1    my_estimator           Z     Bootstrapped SE           <NA> (0.28) (0.14) (0.00)   (0.10)
4         2    my_estimator           Z     Bootstrapped SE           <NA> (0.16) (0.13) (0.08)   (0.08)
  mean_estimate sd_estimate mean_se type_s_rate mean_estimand
1          1.00        0.77    0.69         NaN          1.05
2          0.59        0.86    0.67        0.00          0.80
3        (0.25)      (0.14)  (0.05)        (NA)        (0.11)
4        (0.23)      (0.15)  (0.04)        (NA)        (0.12)
```

Seems the number of group names hard coded as 3 but this should be responsive to the diagnosis.
```
  group_names <- diagnosands_df_names[1:3]
  value_names <- diagnosands_df_names[-(1:3)]
```

eg  Perhaps revert to something like what's in the version in 
`DesignLibrary` for this:

```
  k <- min(which(grepl("se\\(", diagnosands_df_names))) - 2
  group_names <- diagnosands_df_names[1:k]
  value_names <- diagnosands_df_names[-(1:k)]
```

Better solution though would be one with diagnosand labels that requires less interpretation based on (

Note also in printout above:

* column reordering is not sensitive to # columns; design column first makes sense; bes to omit I think
* looks to me like (ses) are not being spliced in below the diagnosands and that there is a repetition of labels on both the diagnosand and se rows

We should be aiming for:

```kable(DesignLibrary::reshape_diagnosis(diagnosis))```

|design_ID |estimator_label |coefficient |estimand_label |bias   |rmse   |power  |coverage |mean_estimate |sd_estimate |mean_se |type_s_rate |mean_estimand |
|:---------|:---------------|:-----------|:--------------|:------|:------|:------|:--------|:-------------|:-----------|:-------|:-----------|:-------------|
|1         |my_estimator    |Z           |ATE            |-0.05  |0.82   |0      |0.9      |1             |0.77        |0.69    |NaN         |1.05          |
|          |                |            |               |(0.28) |(0.14) |(0)    |(0.1)    |(0.25)        |(0.14)      |(0.05)  |(NA)        |(0.11)        |
|2         |my_estimator    |Z           |ATE            |-0.21  |0.65   |0.1    |0.9      |0.59          |0.86        |0.67    |0           |0.8           |
|          |                |            |               |(0.16) |(0.13) |(0.08) |(0.08)   |(0.23)        |(0.15)      |(0.04)  |(NA)        |(0.12)        |

**Pull request closed without merging**